### PR TITLE
Start with the training script.

### PR DESCRIPTION
**Pull request by NicolasNoya**
Did the first script for training and want some feedback in order to continue with this task.

### CODE REVIEW

File: research/20231005_script_train.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    save_interval: int = 100,
+):
+    train_dataloader = DataLoader(dataset=train_dataset, batch_size=batch_size)
+    val_dataloader = DataLoader(dataset=val_dataset, batch_size=batch_size)
+    criterion = nn.CrossEntropyLoss()
+    optimizer = Adam(model.parameters(), lr=learning_rate)
+
+    # Get the number of labels in the total dataset
+    list_labels = train_dataset.dataset.img_labels + val_dataset.dataset.img_labels
+    set_labels = set(list_labels)
```
Comment (j-oflaherty) [inline research/20231005_script_train.py:60]: You shouldn't have labels in your validation dataset that you don't have in your training dataset. How will it learn it?

If you had to add it cause it was failing, it's probably cause your split didn't sample data from one label. With big enough datasets, this won't be a problem, but when your set is small you will find this issue quite often. The solution is called a stratified split, where you mantain the distribution of the original set in your subsets.

Here's an example of how to do it using sklearn and pytorch: https://discuss.pytorch.org/t/how-to-do-a-stratified-split/62290/5

File: research/20231005_script_train.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            loss = criterion(logits, labels)
+            print("The loss is: ", loss)
+            loss.backward()
+            optimizer.step()
+
+            running_loss += loss.item()
+
+            for metric in metrics.keys():
+                # Review later
+                metrics[metric].update(logits, labels)
```
Comment (j-oflaherty) [inline research/20231005_script_train.py:88]: Let's leave this out for the time being, since we aren't doing anything with the metrics yet. You will have time to look into this with this when we start making logs.

**Review by j-oflaherty — APPROVED** [review]
Good job so far. Next up, let's move this research to a train.py script that we can call from the terminal. This way, we can leave a train running outside vscode.

File: pento/train.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    optimizer = Adam(model.parameters(), lr=learning_rate)
+
+    best_loss = float("inf")
+
+    for epoch in range(num_epochs):
+        running_loss = 0
+        for batch, labels in train_dataloader:
+            optimizer.zero_grad()
+
+            logits = model(batch.float())
```
Comment (j-oflaherty) [inline pento/train.py:55]: Why are you converting the input to float? Isn't it already a float?

File: pento/train.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+def save_checkpoint(
+    model: nn.Module, optimizer: Optimizer, epoch: int, lost: float, path: str
+):
+    torch.save(
+        {
+            "epoch": epoch,
+            "model_state_dict": model.state_dict(),
+            "optimizer_state_dict": optimizer.state_dict(),
+            "loss": lost,
```
Comment (j-oflaherty) [inline pento/train.py:85]: Rename lost to loss

File: pento/train.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def load_checkpoint(path: str, model: nn.Module, optimizer: Optimizer):
+    checkpoint = torch.load(path)
+    model.load_state_dict(checkpoint["model_state_dict"])
+    optimizer.load_state_dict(checkpoint["optimizer_state_dict"])
+    epoch = checkpoint["epoch"]
+    loss = checkpoint["loss"]
+
+    return epoch, loss
```
Comment (j-oflaherty) [inline pento/train.py:98]: What is this function for? If it's not going to be used, we can remove it. Else, add the functionality to be able to resume training from the checkpoint.

**Review by j-oflaherty — CHANGES_REQUESTED** [review]

**Review by j-oflaherty — COMMENTED** [review]
Lot's of progress being made. Big PR, more to comments, so try to tackle them one at a time and resolve them as you finish them. Don't hesitate to ping agus or me if you don't understand something about the way to do A or B, since a lot of things aren't easy at first and become clearer with practice.

File: pento/configurationandtrain.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    train(
+        model=model,
+        out_dir=out_dir,
+        train_dataset=data_train,
+        val_dataset=data_val,
+        num_epochs=num_epochs,
+        learning_rate=learning_rate,
+        batch_size=batch_size,
+        save_interval=save_interval,
+    )
```
Comment (j-oflaherty) [inline pento/configurationandtrain.py:118]: This function objective overlaps with the function configurationAndtrain that calls it. It continues loading this from the config and setting up stuff, that can and should be done in the previous function. It leads to a rabbit hole where it's become unclear where setup ends and where train starts. This code is fairly simple, but as it gets more complex it will be harder to keep track of it.

File: Configurtation_files/experiment1.yaml
[the remark this answers is not in the document]
Comment (a-freitas) [reply Configurtation_files/experiment1.yaml:1]: Also, avoid capital letters in general (except for names)

File: Configurtation_files/experiment1.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+model:
+  classifier_neurons: [1000, 1000, 1000]
+  num_classes: 4
+  back_bone: ResNet50_Weights # Future feature
+
+
+dataset_configuration:
+  img_dir: "../data/LittleTrain"
```
[the remark this answers is not in the document]
Comment (a-freitas) [reply Configurtation_files/experiment1.yaml:8]: Another way to avoid conflicts with paths when changing instances is to use a .env file to specify all the paths as environment variables easily.

File: Configurtation_files/experiment1.yaml [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    transform_always_apply:
+      transform1:
+        name: Resize
+        params: 
+          size: [512, 512]
+    transform_sometimes_apply:
+      transform1:
+        name: RandomRotation
+        params: 
+          degrees: 90
```
[the remark this answers is not in the document]
Comment (a-freitas) [reply Configurtation_files/experiment1.yaml:20]: I wouldn't differentiate the augmentations depending on if they are not always applied or not, by name. I would leave that as inferred from the presence of ```probability``` or whichever name you choose.

File: models/dataset.py [hunk 9]
```diff
@@ -37,6 +43,13 @@ def __init__(
         self.transform = transform
         self.target_transform = target_transform
 
+        self.mean_red = red_mean
```
Comment (a-freitas) [inline models/dataset.py:46]: You could use the class wrapper @dataclass to avoid using ```self.x=x``` inside the init

File: pento/data_processing.py
[the remark this answers is not in the document]
Comment (NicolasNoya) [reply pento/data_processing.py:1]: It was given by the Pento's template.
